File: CS_LinkedLists/LinkedLists/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LinkedLists {
    class Stack<Item> {
        // 栈顶结点
        private Node nil = new Node();
        // 元素个数
        private int N;
        // 结点嵌套类
        class Node {
            public Item item;
            public Node next;
        }
        // 判断栈是否为空
        public bool IsEmpty() {
            // 或者: N == 0.
            return nil.next == null; 
        } 

        // 获取栈大小
        public int size() {
            return N;
        }

        // push:Add item to top of stack.
        public void push(Item item) { // 
            Node newNode = new Node();
            newNode.item = item;
            newNode.next = nil.next;

            nil.next = newNode;

            N++;
        }
        // pop:Remove item from top of stack. 
        public Item pop() {
            if (IsEmpty()) {
                Console.WriteLine("Empty Stack");
                return default(Item);
            }
            Item oldFirstItem = nil.next.item;

            nil.next = nil.next.next;

            N--;
            return oldFirstItem;
        }
        // 测试
        public void test() {
            Stack<String> stack = new Stack<String>();
            stack.push("ice");
            stack.push("arl");
            stack.push("111");
            stack.push("222");
            stack.pop();
            while(stack.size() > 0) {
                Console.WriteLine(stack.pop());
            }
        }
    }
}

File: Python/data_structures/stack.py
# 栈的特点是「先进后出」，一般有这几个操作
# push 将一个元素存入栈中 O(1)
# pop  将一个元素从栈中取出，并在栈中删除它O(1)

# top  将一个元素从栈中取出
# is_empty 查看栈是否是空的

# size 栈的长度


class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        return self.items.pop()

    def is_empty(self):
        return self.items == []

    def top(self):
        return self.items[len(self.items) - 1]

    def size(self):
        return len(self.items)